### ISSUE TITLE
Deep nested validation broken

### ISSUE THREAD

**Issue by hood** [typestack/class-validator#614]
### Description
When trying to validate nested objects using `@ValidateNested()`,  `whitelist: true` and `forbidNonWhitelisted: true`, error ` "property XY should not exist"` gets thrown for no apparent reason.

### Reproduction
Here a vastly simplified version of my use case:

```
class ProductPricing {
    @Type(() => ProductPricingOptionsDTO)
    options: ProductPricingOptionsDTO;
}
```
```
class ProductPricingOptionsDTO{
    @ValidateNested()
    @Type(() => OnlineUserPricingOptionsDTO)
    onlineUserPrices: OnlineUserPricingOptionsDTO;
}
```
```
class OnlineUserPricingOptionsDTO{
    @IsNumber()
    @Min(2)
    price: number;
}
```

and validating the following payload:
```
{
    "options": {
        "onlineUserPrices": {
            "price": 2
        }
    }
}
```
    
with ```whitelist: true, forbidNonWhitelisted: true``` throws an error containing the following message:
```
"constraints": {
    "whitelistValidation": "property price should not exist"            }
}
```

### Environment

- [x] nodejs: 
- [ ] browser: 
- [x] typeorm: 

**class-validator version:** <code>^0.12.1</code>

**Comment by asomethings** [comment]
Experiencing same issue. Did you manage to fix this issue?

**Comment by Yokutto** [comment]
Setting only `forbidNonWhitelisted: true` is necessary to reproduce this problem, that will throw the `property X should not exist` message

For setting only `forbidUnknownValues: true`, the class-validator will throw with `an unknown value was passed to the validate`

If both options are enabled, it will give an error in the first check rule

For example:
If my PipeValidation is 
```typescript
new ValidationPipe({
   whitelist: true,
   forbidUnknownValues: true,
   forbidNonWhitelisted: true,
})
```
It will throw the `forbidUnknownValues` message because it's the first rule checked at ValidationPipe, the same occurs if `forbidNonWhitelisted` is above `forbidUnknown` but it will throw `property X should not exist` message instead

**Comment by rbiggers** [comment]
Any updates on this?

**Comment by murbanowicz** [comment]
Just hit this issue...

**Comment by gabrieloczkowski** [comment]
Hi!
I had the same issue but I managed to resolve it by using:
```
transformOptions: {
    enableImplicitConversion: true,
},
```
Please let me know if it helped also in your cases.

**Comment by murbanowicz** [comment]
It works in terms of passing the validation, but... It does not validate the nested objects at all now :/ 

```typescript
@IsNotEmpty()
  @IsArray()
  @ArrayNotEmpty()
  @ValidateNested({ each: true })
myNestedObjects!: NestedObjectType[]
```

```typescript
export class NestedObjectType{
  @IsNotEmpty()
  @IsString()
  firstName!: string;

  @IsNotEmpty()
  @IsString()
  lastName!: string; 
}
```

I can miss fields or pass wrong values - validation still passes without any issue :/

**Comment by gabrieloczkowski** [comment]
I tested such example
```
export class Address {
    @IsString()
    public street!: string;
}

export class CreateHippoParams {
    @IsString()
    public name!: string;

    @IsIn(Object.values(Gender))
    @ApiProperty({ enum: Gender, enumName: GENDER_ENUM_NAME })
    public gender!: Gender;

    @IsPositive()
    public age!: number;

    @ValidateNested()
    @Type(() => Address)
    public address!: Address;
}
```
And it validated `address` correctly for optimistic/pessimistic paths.

**Comment by murbanowicz** [comment]
Cheers! We have found out that adding `@Type(() => NestedType)` was the thing. It works without adding that parameter which you suggested.
Adding comment for "future generations".

**Comment by gabrieloczkowski** [comment]
Ok. Update on my case. `@Type` decorator didn't work as should when used `yarn link` along with `class-transformer` in peerDependency. So option `enableImplicitConversion: true` helped me because of not working `@Type` but when not linked everything works as it should 👍 .

**Comment by rbiggers** [comment]
@murbanowicz adding ```@Type(() => NestedType)``` seems to only be working on second level objects. If you go down one more level like in the OP validation doesn't work on the third level object. I will try @gabrieloczkowski suggestion for ```enableImplicitConversion: true``` today and see if it works.

**Comment by rbiggers** [comment]
@gabrieloczkowski ```enableImplicitConversion: true``` did not work for me. Could you share an example of your implementation?

**Comment by gabrieloczkowski** [comment]
> @gabrieloczkowski `enableImplicitConversion: true` did not work for me. Could you share an example of your implementation?

@rbiggers 
https://github.com/typestack/class-validator/issues/614#issuecomment-698336485
```
{
    whitelist: true,
    forbidNonWhitelisted: true,
    forbidUnknownValues: true,
    transform: true,
    transformOptions: {
        enableImplicitConversion: true,
    },
}
```

**Comment by jpabeem** [comment]
@gabrieloczkowski Where did you enable these attributes? I can only find the first three attributes in the configuration when validating with `validate()` of `class-validator`.

**Comment by asomethings** [comment]
@jpabeem configurations which @gabrieloczkowski wrote are for`ValidationPipe` which is used in [nestjs](https://github.com/nestjs/nest)

**Comment by jpabeem** [comment]
Thanks! Makes sense. Hope we can get support for deeply-nested validation in a future CR :-)

**Comment by stephen-dahl** [comment]
I am running

```
    new ValidationPipe({
      whitelist: true,
      transform: true,
      forbidNonWhitelisted: true,
      transformOptions: {
        enableImplicitConversion: true,
      },
    }),
```

and everything works except enforcing the whitelist on sub-objects

**Comment by MickL** [comment]
Is this targeted to be fixed? The issue is open for a year now :(

**Comment by MickL** [comment]
@stephen-dahl for using `enableImplicitConversion ` the documentation says:
> NOTE If you use class-validator together with class-transformer you propably DON'T want to enable this function.

Is this targeted to be fixed? The issue is open for a year now :(

**Comment by stephen-dahl** [comment]
> @stephen-dahl for using `enableImplicitConversion ` the documentation says:
> 
> > NOTE If you use class-validator together with class-transformer you propably DON'T want to enable this function.
> 
> Isnt this issue related to class-transformer instead of class-validator? Because when you you class-transformer your sub-object needs to be transformed to a class, too. This is done by using `@Type(() => Photo)`. So actually there is no issue here and this can be closed, or am I wrong?

@MMy object is being cast to type photo and all its keys are being cast to the appropriate types. My issue here is that keys that are not on my Photo object are not throwing an error saying that they should not exist, but they are on the output object after validation.
I take that note as a general recommendation that some validators are less usefull after casting but not completely. for example IsString on a field that is cast to string will always get a string or undefined so it basically just verifies it is not undefined

**Comment by MickL** [comment]
I am pretty sure this issue can be **closed** because nested validation is NOT broken:

When using `@ValidateNested()` from class-validator then the property needs to be an instance of the desired class not a plain object. If you use class-transformer then it will not automatically transform your nested objects into classes. Instead you need to add the decorator `@Type(() => MyClass)` so class-transformer will transform your nested objects into instances of the class and THEN `@ValidateNested()` from class-validator will work as expected.

If you activate `enableImplicitConversion` then class-validator will do this transformation for you, but as the documentation states you should **not** enable `enableImplicitConversion` when using class-transformer.

@hood May you verify and close this issue?

**Comment by rbiggers** [comment]
@MickL are you sure it is not still broken? Have you tried to validate objects that are nested more than 2 levels deep? That is the issue i was running into. It works on second level objects but fails for 3rd level or deeper. Even when adding @Type(() => MyClass)

I don't think this issue should be closed until it is confirmed the validation works on 3rd level objects.

**Comment by MickL** [comment]
I didnt check multi-level but i would first check if this is an issue with class-transformer. Use console.log and check if all objects are actually a class. If yes then this issue should stay open, if no then please close this issue but open a new one at class-transformer repo.

**Comment by rbiggers** [comment]
That sounds great I will leave that up to @hood as it is his original ticket and I currently don't have the bandwidth to support. For my work around I just made a custom validator but now that I'm actually looking at the code you may be right about it being a class-transformer issue... I am having to use `plainToClass()` to get the Class before calling `validate()`

**Comment by luukvhoudt** [comment]
I also encountered this issue when using the `plainToClass()` class transformer. I tried to reproduce this issue in this [CodeSandbox](https://codesandbox.io/s/nested-validation-test-yzqio). While doing so I noticed that `plainToClass()` doesn't convert nested "class" objects to the desired class, what actually shouldn't be an issue if you're **not** strictly validating the type of the nested object (like in my example).

As you might have seen my example includes some tests what, I think, cover all cases regarding this issue and if you run them you will see that they all pass. Which is strange because when running the server, defined in `index.ts`, you can see that it does give an error. The server runs a validation for data with and without class transformer and while doing so the `whitelist` and `forbidNonWhitelisted` flag are set to `true`, resulting in errors for the validation with class transformer. This is the only way I could manage to make the validator return validation errors.

To summarize I think that this issue can be solved by making the `whitelist` function declared [here](https://github.com/typestack/class-validator/blob/master/src/validation/ValidationExecutor.ts#L114) recursive, for objects and arrays of objects. Thus for now I would suggest you to convert all your nested objects to the desired class. If you're using a class transformer then this could potentially solve the issue.

**Comment by maaaaagda** [comment]
I had `property xxx should not exist` with whitelist rule set to true error but my problem was that I wasn't specifying any decorators for nested properties, whereas in the docs they write:

_(validate)... will strip all properties that don't have any decorators. If no other decorator is suitable for your property, you can use @Allow decorator_

Now everything works fine with @ValidateNested

**Comment by barbozafernando** [comment]
> I had `property xxx should not exist` with whitelist rule set to true error but my problem was that I wasn't specifying any decorators for nested properties, whereas in the docs they write:
> 
> _(validate)... will strip all properties that don't have any decorators. If no other decorator is suitable for your property, you can use @Allow decorator_
> 
> Now everything works fine with @ValidateNested

This solved my problem. Thank you.

**Comment by braaar** [comment]
Here is a runnable version of the original example from @hood

You will want to place this code inside a function if you are not using ESM modules, due to the top level `await`.

I ran this using class-validator 0.13.2.

```ts
import { Type, plainToClass } from 'class-transformer';
import { IsNumber, Min, ValidateNested, validate } from 'class-validator';

class OnlineUserPricingOptionsDTO {
  @IsNumber()
  @Min(2)
  price: number;
}

class ProductPricingOptionsDTO {
  @ValidateNested()
  @Type(() => OnlineUserPricingOptionsDTO)
  onlineUserPrices: OnlineUserPricingOptionsDTO;
}
class ProductPricing {
  @Type(() => ProductPricingOptionsDTO)
  options: ProductPricingOptionsDTO;
}

const result = await validate(plainToClass(ProductPricing, payload), {
  whitelist: true,
  forbidNonWhitelisted: true,
});

console.log(result);
```

This result in the following console message:
```
[
  ValidationError {
    target: ProductPricing { options: [ProductPricingOptionsDTO] },
    value: ProductPricingOptionsDTO {
      onlineUserPrices: [OnlineUserPricingOptionsDTO]
    },
    property: 'options',
    children: undefined,
    constraints: { whitelistValidation: 'property options should not exist' }
  }
]
```

As we can see, the `options` property does not have a class-validator decorator on it, so in order to pass validation with whitelisting, we must either use the `@Allow` decorator (as others have stated), or use an actual validation decorator we want to validate against, which in this case could be `@ValidateNested`. 

When we add such a decorator to the `options` field, validation passes.

I think this issue has been sufficiently discussed and the original reproduction has been explained and addressed. Thanks for your explanations!
